### DISCUSSION TITLE
What's the idiomatic way to show toast / success messages on form submission?

### DISCUSSION THREAD

**Discussion by tmcw** [edmundhung/conform#624] [Q&A]
This is a little tricky with Remix and `useFetcher`, but I think the norm is to use a `useEffect` and look for `fetcher.state` and `fetcher.data`. `useForm` in Conform seems to manage the fetcher behind the scenes. Is there a way to get its fetcher, or a callback, or another accessible way to know when a form submission has succeeded?

**Comment by edmundhung** [comment]
Hi @tmcw, Conform is not integrated with Remix. All it does is to call `event.preventDefault()` for you whenever appropriate. If you setup the form with the Remix Form component, you should be able to get the state with `useNavigation()` from remix or if you are using the `fetcher.Form` component, then you should also be able to access the state with `fetcher.state` by yourself.

**Comment by Pouet--** [comment]
This a not a Conform-related question but I ran into a similar case. I previously used useEffect to watch for the server response and display my toast.

But now I can trigger the toast directly from the loader and action functions, which is a great improvement.

I used this epic-stack example : https://github.com/epicweb-dev/epic-stack/blob/main/docs/toasts.md

Look inside the toast.server file he is talking about : https://github.com/epicweb-dev/epic-stack/blob/main/app/utils/toast.server.ts
And update your root file to trigger toasts : https://github.com/epicweb-dev/epic-stack/blob/main/app/root.tsx
